Duplicate Key Violation When Reauthenticating After Account Deletion via Google

### Before reporting an issue

- [x] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

authentication

### Describe the bug

**Issue Description:**  
We encountered an issue in Keycloak **26.1.4** where users cannot immediately reauthenticate after deleting their account when using Google (or another external identity provider). The error message indicates a **duplicate key violation** in the database.  


### Version

26.1.4

### Regression

- [ ] The issue is a regression

### Expected behavior

### 🔍 **Expected Behavior:**  
The user should be able to reauthenticate successfully as if they are logging in for the first time.  

### Actual behavior

### 🔄 **Steps to Reproduce:**  
1. Configure authentication via **Google** (or another external provider).  
2. Navigate to **Authentication → Required Actions** and enable **"Delete Account"**.  
3. Go to **Realm Settings → User Registration → Default Roles**, add the **"Delete Account"** role.  
4. Log in with **Google**, then delete the account through the UI.  
5. Immediately attempt to log in again with Google.  
6. The following error occurs:  

```
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "constraint_offl_us_ses_pk2"
Detail: Key (user_session_id, offline_flag)=(4eb36d91-6d49-442f-8fad-9be63896d89b, 0) already exists.
```

### How to Reproduce?

### 🔄 **Steps to Reproduce:**  
1. Configure authentication via **Google** (or another external provider).  
2. Navigate to **Authentication → Required Actions** and enable **"Delete Account"**.  
3. Go to **Realm Settings → User Registration → Default Roles**, add the **"Delete Account"** role.  
4. Log in with **Google**, then delete the account through the UI.  
5. Immediately attempt to log in again with Google.  
6. The following error occurs:  

```
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "constraint_offl_us_ses_pk2"
Detail: Key (user_session_id, offline_flag)=(4eb36d91-6d49-442f-8fad-9be63896d89b, 0) already exists.
```


### Anything else?

Keycloak Version: 26.1.4
Database: PostgreSQL

⚠ Important Note:
The issue resolves itself after approximately 5 minutes, allowing the user to log in successfully again. However, the immediate failure creates a poor user experience and may indicate an underlying issue with session or user data cleanup.

❓ Possible Cause:
It seems that after account deletion, some session-related data (e.g., user_session_id) is not fully removed from the database. When the user tries to log in again immediately, Keycloak attempts to reuse an existing session, resulting in a constraint violation in offline_user_session.

We would appreciate any insights or a potential fix for this issue! 🙏

--- a/services/src/main/java/org/keycloak/authentication/requiredactions/DeleteAccount.java
+++ b/services/src/main/java/org/keycloak/authentication/requiredactions/DeleteAccount.java
@@ -42,7 +42,9 @@ import org.keycloak.models.RoleModel;
 import org.keycloak.models.UserManager;
 import org.keycloak.models.UserModel;
 import org.keycloak.services.managers.AuthenticationManager;
+import org.keycloak.services.managers.AuthenticationSessionManager;
 import org.keycloak.services.messages.Messages;
+import org.keycloak.sessions.AuthenticationSessionModel;
 
 public class DeleteAccount implements RequiredActionProvider, RequiredActionFactory {
 
@@ -94,7 +96,7 @@ public class DeleteAccount implements RequiredActionProvider, RequiredActionFact
             .detail(Details.USERNAME, user.getUsername())
             .success();
 
-        cleanSession(context, RequiredActionContext.KcActionStatus.SUCCESS);
+        removeAuthenticationSession(context, session);
 
         context.challenge(context.form()
             .setAttribute("messageHeader", "")
@@ -184,5 +186,10 @@ public class DeleteAccount implements RequiredActionProvider, RequiredActionFact
   @Override
   public int getMaxAuthAge() {
     return 0;
+  }
+
+  private void removeAuthenticationSession(RequiredActionContext context, KeycloakSession session) {
+    AuthenticationSessionModel authSession = context.getAuthenticationSession();
+    new AuthenticationSessionManager(session).removeAuthenticationSession(authSession.getRealm(), authSession, true);
   }
 }

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/actions/DeleteAccountActionTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/actions/DeleteAccountActionTest.java
@@ -11,6 +11,7 @@ import org.junit.Rule;
 import org.junit.Test;
 import org.keycloak.admin.client.resource.UserResource;
 import org.keycloak.authentication.requiredactions.DeleteAccount;
+import org.keycloak.cookie.CookieType;
 import org.keycloak.events.EventType;
 import org.keycloak.models.AccountRoles;
 import org.keycloak.representations.idm.ClientRepresentation;
@@ -25,6 +26,7 @@ import org.keycloak.testsuite.auth.page.login.DeleteAccountActionConfirmPage;
 import org.keycloak.testsuite.pages.ErrorPage;
 import org.keycloak.testsuite.pages.LoginPage;
 import org.keycloak.testsuite.util.UserBuilder;
+import org.openqa.selenium.Cookie;
 
 public class DeleteAccountActionTest extends AbstractTestRealmKeycloakTest {
 
@@ -73,6 +75,44 @@ public class DeleteAccountActionTest extends AbstractTestRealmKeycloakTest {
     Assert.assertEquals(users.size(), 0);
   }
 
+    @Test
+    public void testReauthenticateAfterDeletingAccount() {
+        loginPage.open();
+
+        UserRepresentation userRep = UserBuilder.create()
+                .username("delete-user")
+                .password("password")
+                .enabled(true)
+                .requiredAction(DeleteAccount.PROVIDER_ID)
+                .build();
+        testRealm().users().create(userRep).close();
+        addDeleteAccountRoleToUserClientRoles(userRep.getUsername());
+
+        loginPage.login(userRep.getUsername(), "password");
+
+        Assert.assertTrue(deleteAccountPage.isCurrent());
+
+        Cookie authSessionCookie = driver.manage().getCookieNamed(CookieType.AUTH_SESSION_ID.getName());
+        deleteAccountPage.clickConfirmAction();
+
+        events.expect(EventType.DELETE_ACCOUNT);
+
+        List<UserRepresentation> users = testRealm().users().search(userRep.getUsername());
+
+        Assert.assertEquals(users.size(), 0);
+
+        testRealm().users().create(userRep).close();
+        addDeleteAccountRoleToUserClientRoles(userRep.getUsername());
+        loginPage.open();
+        Cookie newAuthSessionCookie = driver.manage().getCookieNamed(CookieType.AUTH_SESSION_ID.getName());
+        Assert.assertFalse(authSessionCookie.getValue().equals(newAuthSessionCookie.getValue()));
+        loginPage.login(userRep.getUsername(), "password");
+        Assert.assertTrue(deleteAccountPage.isCurrent());
+        deleteAccountPage.clickConfirmAction();
+        users = testRealm().users().search(userRep.getUsername());
+        Assert.assertEquals(users.size(), 0);
+    }
+
   @Test
   public void deleteAccountFailsWithoutRoleFails() {
     removeDeleteAccountRoleFromUserClientRoles();
@@ -87,7 +127,11 @@ public class DeleteAccountActionTest extends AbstractTestRealmKeycloakTest {
 
 
   private void addDeleteAccountRoleToUserClientRoles() {
-    UserRepresentation user = ActionUtil.findUserWithAdminClient(adminClient, "test-user@localhost");
+    addDeleteAccountRoleToUserClientRoles("test-user@localhost");
+  }
+
+  private void addDeleteAccountRoleToUserClientRoles(String username) {
+    UserRepresentation user = ActionUtil.findUserWithAdminClient(adminClient, username);
     ApiUtil.assignClientRoles(adminClient.realm("test"), user.getId(), "account", AccountRoles.DELETE_ACCOUNT);
   }
 

